### PR TITLE
staging: run recipes from firebase

### LINKED ISSUE #167: Run a recipe from firebase
## Use Case
Want to be able to request a recipe and run it from firebase using `pack -r firebase:[RECIPE_NAME] -c [CONFIG_FILE]`

## Solution
In autopack/__init__: `load_file`: add code to the TODO section to download the data and return it by calling into the DB Handler. 
This will involve following references and downloading the data for each reference. 

1. request the recipe which will have references to a composition. 
2. create an `objects` dict that contains all the objects contained in the recipe with the keys being the name of the obj
3. update the `composition` dict to have the `object` pointers point to the keys in the `objects` dict 

## Alternatives

In recipe loader. _read you could request and make the objects dict here instead. 
pseudo code:

        for comp in composition:
            walk down tree:
                if object_ref is to firebase:
                    if not r"objects" in recipe_data:
                        create recipe_data["objects"]
                    download object, place in recipe_data["objects"]
         

example factoring that may be useful
def walk_through_ composition_and_do_something(callback): 
   for comp in composition:
      for region etc:
          callback(data)

remember to look for ways to factor out code that is duplicated.

### PR DESCRIPTION
**Pull request by rugeli**
Problem
=======
What is the problem this work solves, including
closes #167 

Solution
========
This is a staging branch for run recipes from firebase.

features in this branch that need to be reviewed:
- a new DBUploader class for data conversion and uploading recipes into modules

already reviewed: 
- firebase credential input
- AWSHandler for storing simularium results to S3
- save metadata to firebase 
- open a new tab for simularium result 
- handlers are initiated using the DATABASE_IDS enum

## Type of change
Please delete options that are not relevant.

* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Have firebase project setup and your personal credentials ready
2. Upload a recipe to your firebase: `upload -r examples/recipes/v2/gradients.json`
3. Pack the remote recipe you uploaded: `pack -r firebase:recipes/gradients_v_default -c examples/packing-configs/run.json`
4. check up the result on Simularium, AWS S3 and Firebase

Keyfiles (delete if not relevant):
-----------------------
1. `DBRecipeHandler.py`
2. `autopack/__init__.py`

### CODE REVIEW

**Comment by rugeli** [comment]
passed the following tests:
- uploaded a new local recipe to firebase: upload accepted 
- uploaded a recipe that exists in firebase: upload rejected   
- packed a local recipe
- packed a firebase recipe

current issue:
- the analyze workflow lacks a mechanism to input firebase credentials. (still exploring options)

File: cellpack/autopack/loaders/recipe_loader.py [hunk 1]
```diff
@@ -165,8 +166,102 @@ def _migrate_version(self, old_recipe):
...
+                                    object_copy, objects, gradients
+                                )
+                        else:
+                            composition[comp_name]["regions"][region_name].append(
+                                region_item["name"]
+                            )
+        return objects, gradients, composition
+
+    @staticmethod
+    def _compile_recipe_from_firebase(db_recipe_data, obj_dict, grad_dict, comp_dict):
```
Comment (meganrm) [inline cellpack/autopack/loaders/recipe_loader.py:238]: Shouldn't have anything referencing firebase in recipe loader. this should be in the database handler.

File: cellpack/autopack/loaders/recipe_loader.py [hunk 2]
```diff
@@ -26,10 +26,11 @@ class RecipeLoader(object):
     # TODO: add all default values here
     default_values = default_recipe_values.copy()
 
-    def __init__(self, input_file_path, save_converted_recipe=False):
+    def __init__(self, input_file_path, db_handler=None, save_converted_recipe=False):
```
Comment (meganrm) [inline cellpack/autopack/loaders/recipe_loader.py:29]: Should probably not be sending a db_handler here? Maybe send it to auto pack. Should also be optional and default to None so when we run our tests and when we run a recipe locally it doesn't have any firebase stuff to do

File: cellpack/bin/pack.py [hunk 3]
```diff
@@ -20,18 +23,29 @@
 ###############################################################################
 
 
-def pack(recipe, config_path=None, analysis_config_path=None):
+class DATABASE_IDS(Enum):
+    FIREBASE = "firebase"
+    GITHUB = "github"
```
Comment (meganrm) [inline cellpack/bin/pack.py:28]: this should be in the db handler file or in `interface` which is where we store data objects/enums

**Comment by rugeli** [comment]
Work update:
- refactored functions responsible for remote recipe downloading from `recipe_loader` into `DBRecipeLoader` class in `DBRecipeHandler.py`
- moved the `database_ids` enum to `interface_objects`
- removed the `db_handler` parameter from both `pack` and `recipe_loader`
- initialized `db_handler` and `FirebaseHandler()` in `autopack/__init__.py` (subject to feedback and discussion)

**Comment by mogres** [comment]
Can you add instructions to create firebase credentials somewhere? These could be a page in the [wiki](https://github.com/mesoscope/cellpack/wiki) that can be linked elsewhere.

**Comment by rugeli** [comment]
> Can you add instructions to create firebase credentials somewhere? These could be a page in the [wiki](https://github.com/mesoscope/cellpack/wiki) that can be linked elsewhere.

The updated README is in the `upload local recipes to s3`PR - #189, you can view the instructions here: https://github.com/mesoscope/cellpack/pull/189/files

So after some discussion and having addressed the changes, the recommended PR merge order should be: #191 - #189 - (WIP) run recipes with "inherit" key - #179, feel free to review the first two when you have time. I'll convert this PR to a draft for now. Sorry for the confusion!

**Comment by mogres** [comment]
> > Can you add instructions to create firebase credentials somewhere? These could be a page in the [wiki](https://github.com/mesoscope/cellpack/wiki) that can be linked elsewhere.
> 
> The updated README is in the `upload local recipes to s3`PR - #189, you can view the instructions here: https://github.com/mesoscope/cellpack/pull/189/files
> 
> So after some discussion and having addressed the changes, the recommended PR merge order should be: #191 - #189 - (WIP) run recipes with "inherit" key - #179, feel free to review the first two when you have time. I'll convert this PR to a draft for now. Sorry for the confusion!

Thanks for the update! I'll follow the recommended order.

File: cellpack/tests/test_data_doc.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        "partners": {"all_partners": []},
+    },
+    "name": "nucleus",
+}
+
+
+def test_is_nested_list():
+    assert DataDoc.is_nested_list([]) is False
+    assert DataDoc.is_nested_list([[], []]) is True
+    assert DataDoc.is_nested_list([[1, 2], [3, 4]]) is True
```
Comment (mogres) [inline cellpack/tests/test_data_doc.py:31]: Should `DataDoc.is_nested_list([1, [1, 2]])` be `True` or `False`?

File: .gitignore [hunk 5]
```diff
@@ -88,6 +88,7 @@ dask-worker-space
 
 # dotenv
 .env
+.creds
```
Comment (mogres) [inline .gitignore:91]: This is already added at the end of the `.gitignore` file
```suggestion
```

**Review by mogres — COMMENTED** [review]
> > Can you add instructions to create firebase credentials somewhere? These could be a page in the [wiki](https://github.com/mesoscope/cellpack/wiki) that can be linked elsewhere.

The updated README is in the `upload local recipes to s3`PR - #189, you can view the instructions here: https://github.com/mesoscope/cellpack/pull/189/files

So after some discussion and having addressed the changes, the recommended PR merge order should be: #191 - #189 - (WIP) run recipes with "inherit" key - #179, feel free to review the first two when you have time. I'll convert this PR to a draft for now. Sorry for the confusion!   

Maybe we could also add the link to create firebase credentials in the prompt to provide the credentials path?

**Comment by mogres** [comment]
Looks good overall! I was able to upload and run the gradient recipe remotely. Thanks for adding detailed tests and documentation :)

**Review by mogres — APPROVED** [review]

**Comment by rugeli** [comment]
> > > Can you add instructions to create firebase credentials somewhere? These could be a page in the [wiki](https://github.com/mesoscope/cellpack/wiki) that can be linked elsewhere.
> 
> The updated README is in the `upload local recipes to s3`PR - #189, you can view the instructions here: https://github.com/mesoscope/cellpack/pull/189/files
> 
> So after some discussion and having addressed the changes, the recommended PR merge order should be: #191 - #189 - (WIP) run recipes with "inherit" key - #179, feel free to review the first two when you have time. I'll convert this PR to a draft for now. Sorry for the confusion!
> 
> Maybe we could also add the link to create firebase credentials in the prompt to provide the credentials path?

Good idea! We are going to refactor the the authentication methods to integrate firebase credentials in issue #214

File: cellpack/autopack/loaders/recipe_loader.py [hunk 6]
```diff
@@ -177,24 +185,36 @@ def _read(self):
...
-                    obj["partners"] = Partners(partner_settings)
+                # the key "all_partners" exists in obj["partners"] if the recipe is downloaded from a remote db
+                partner_settings = (
+                    []
+                    if (
+                        "partners" in obj
+                        and "all_partners" in obj["partners"]
+                        and not obj["partners"]["all_partners"]
+                    )
+                    else obj.get("partners", [])
```
[the remark this answers is not in the document]
Comment (rugeli) [reply cellpack/autopack/loaders/recipe_loader.py:207]: Thank you for asking this! The if statement here is to determine whether the recipes retrieved from firebase have partners. During the upload process, the key "all_partners" is added to "partners" of each obj no matter there is partner or not. 

The structure of the partners in Firebase is as follows:
- for recipes without partners: `obj[partners] == {'all_partners': []}`
- for recipes with partners:  `obj[partners] == {'all_partners': [<cellpack.autopack.interface_objects.partners.Partner object at xyz>]} `

File: cellpack/autopack/loaders/recipe_loader.py [hunk 7]
```diff
@@ -177,24 +185,36 @@ def _read(self):
...
+                )
+                obj["partners"] = Partners(partner_settings)
                 if "type" in obj and not INGREDIENT_TYPE.is_member(obj["type"]):
                     raise TypeError(f"{obj['type']} is not an allowed type")
 
         # handle gradients
-        if "gradients" in recipe_data:
+        # gradients in firebase recipes are already stored as a list of dicts
+        if "gradients" in recipe_data and not isinstance(
+            recipe_data["gradients"], list
```
[the remark this answers is not in the document]
Comment (rugeli) [reply cellpack/autopack/loaders/recipe_loader.py:216]: Just like with `Partners`, the cellpack system organizes `gradients` into a list during the upload process. Therefore, we can skip processing gradients in firebase recipes.

File: cellpack/tests/test_data_doc.py (hunk 4 above)
Comment (rugeli) [reply cellpack/tests/test_data_doc.py:31]: Good thought! Currently it returns False, though it should return True. While we don't typically encounter such nested lists like [1, [1, 2]] in our recipe data, but you are right that our method should be robust enough to handle all types of nested lists. I just modified the function to run a loop, should be good this time.

**Review by rugeli — COMMENTED** [review]
@mogres Thank you for your thorough review! I've just addressed several issues. Please feel free to point out any additional edge cases or concerns that I may overlooked.

File: cellpack/autopack/loaders/utils.py [hunk 8]
```diff
@@ -21,3 +22,16 @@ def create_output_dir(out_base_folder, recipe_name, sub_dir=None):
...
+
+
+def read_json_file(path):
+    if not Path(path).exists():
+        return None
+    with open(path, "r") as file_name:
+        return json.load(file_name)
+
+
+def write_json_file(path, data):
```
[the remark this answers is not in the document]
Comment (rugeli) [reply cellpack/autopack/loaders/utils.py:34]: I just found out that we have checks for existing keys and contents(like user and firebase creds) before calling this function. So it should be safe to leave this method in its original form.

**Review by meganrm — APPROVED** [review]

**Pull request merged**